### PR TITLE
Add echo_tag_syntax rule

### LINKED ISSUE #2990: Normalizing simple echo tags (short and long forms)
We currently have the `no_short_echo_tag` to expand code from `<?=` to `<?php echo`.

#2520 was an attempt to do the reverse, but I think it has been closed mainly because of lack of interest.

I can take care of this, by implementing to a configurable fixer that can merge the two fixers (`no_short_echo_tag` and `opening_tag_plus_echo_to_short_echo_tag`), into a new `short_echo_tag` fixer.

This fixer could have
- a `format` option (required), with a value of:
  - `'short'` to convert `<?php echo` and `<?php print` to `<?=`
  - `'long'` to convert `<?=` to `<?php echo` or `<?php print`
- a `printOrEcho` option (optional, defaulting to `'echo'`), with a value of:
  - `'echo'` to convert `<?=` to `<?php echo`
  - `'print'` to convert `<?=` to `<?php print`

We could also add more functionalities to this fixer, but those would maybe be better implemented in another fixer. I'm thinking about normalizing spaces and trailing semi-colons (eg formatting `<?php echo 2;?>` as `<?= 2 ?>` or `<?=2?>` or whatever).

Thoughts?

### PR DESCRIPTION
**Pull request by mlocati**
We have a fixer (`no_short_echo_tags`) that expands short echo tags (`<?= $foo ?>`) to long echo tags (`<?php echo $foo ?>`).

What about adding a new fixer (`short_echo_tags`) that can also do the reverse?

Fix #2990

### CODE REVIEW

File: src/Fixer/PhpTag/ShortEchoTagFixer.php [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function createConfigurationDefinition()
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder(self::OPTION_FORMAT, 'How the fixer should process short/long echo tags'))
+                ->setAllowedValues($this->supportedFormatOptions)
+                ->setDefault(self::FORMAT_SHORT)
```
Comment (kubawerlos) [inline src/Fixer/PhpTag/ShortEchoTagFixer.php]: Why not `self::FORMAT_SHORT`?

Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: > Why not `self::FORMAT_SHORT`?

@kubawerlos I'm not sure I understand... it *is* `self::FORMAT_SHORT`...

Comment (kubawerlos) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: @mlocati yes, sorry, I meant why not `self::FORMAT_LONG` - it would be consistent with `no_short_echo_tag`

Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: @kubawerlos I agree: see 451129f30ac68adfa68723b156872b18c5293b14

**Comment by Big-Shark** [comment]
@kubawerlos Any update about this?

File: README.rst [hunk 2]
```diff
@@ -1711,6 +1712,20 @@ Choose from the list of available rules:
...
   *Risky rule: risky when the ``settype`` function is overridden or when used as the 2nd or 3rd expression in a ``for`` loop .*
 
+* **short_echo_tag** [@PhpCsFixer]
+
+  Replace short-echo ``<?=`` with long format ``<?php echo``/``<?php print``
+  syntax, or vice-versa.
+
+  Configuration options:
+
+  - ``format`` (``'long'``, ``'short'``): how the fixer should process short/long echo
```
Comment (kubawerlos) [inline README.rst]: The description could be more clear, like from `no_mixed_echo_print` - `the desired language construct`.

File: dev-tools/trigger-website.sh [hunk 3]
```diff
@@ -4,7 +4,7 @@ set -eu
 TOKEN=$1
 MSG_SUFFIX=${2:-''}
 
-if [ ! -z "$MSG_SUFFIX" ]
+if [ -n "$MSG_SUFFIX" ]
```
Comment (kubawerlos) [inline dev-tools/trigger-website.sh]: Can we revert this change?

File: src/Fixer/PhpTag/NoShortEchoTagFixer.php [hunk 4]
```diff
@@ -12,16 +12,17 @@
...
+use PhpCsFixer\Fixer\DeprecatedFixerInterface;
 use PhpCsFixer\FixerDefinition\CodeSample;
 use PhpCsFixer\FixerDefinition\FixerDefinition;
-use PhpCsFixer\Tokenizer\Token;
-use PhpCsFixer\Tokenizer\Tokens;
 
 /**
  * @author Vincent Klaiber <hello@vinkla.com>
+ *
+ * @deprecated in 2.16, proxy to ShortEchoTagFixer
```
Comment (kubawerlos) [inline src/Fixer/PhpTag/NoShortEchoTagFixer.php]: Version is now out of date, I don't think we need it at all.

File: src/Fixer/PhpTag/ShortEchoTagFixer.php [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        if (self::FORMAT_SHORT === $this->configuration[self::OPTION_FORMAT]) {
+            if (null !== $tokens->findSequence([[\T_OPEN_TAG], [\T_ECHO]])) {
```
Comment (kubawerlos) [inline src/Fixer/PhpTag/ShortEchoTagFixer.php]: In [Cookbook](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.16.0/doc/COOKBOOK-FIXERS.md) the method `isCandidate` has mentioning that it "has to be ultra cheap to execute" - I'm not sure if `findSequence` is fine to have here - @keradus, @SpacePossum, @julienfalque WDYT?

File: src/Fixer/PhpTag/ShortEchoTagFixer.php [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(SplFileInfo $file, Tokens $tokens)
+    {
+        if (self::FORMAT_SHORT === $this->configuration[self::OPTION_FORMAT]) {
+            $this->longToShort($file, $tokens);
+        } else {
```
Comment (kubawerlos) [inline src/Fixer/PhpTag/ShortEchoTagFixer.php]: There is no point to pass `$file` in above and below lines.

File: src/Fixer/PhpTag/ShortEchoTagFixer.php [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+            $this->shortToLong($file, $tokens);
+        }
+    }
+
+    private function longToShort(SplFileInfo $file, Tokens $tokens)
+    {
+        $offset = 0;
+        for (;;) {
+            $foundEcho = $tokens->findSequence([[\T_OPEN_TAG], [\T_ECHO]], $offset);
+            $foundPrint = $tokens->findSequence([[\T_OPEN_TAG], [\T_PRINT]], $offset);
```
Comment (kubawerlos) [inline src/Fixer/PhpTag/ShortEchoTagFixer.php]: Searching bor both and lower using only one - seems like waste of CPU.
Why not with classic `for` loop:
```
for ($i = 0 ; $i < count; $i++)
    if $i is not T_OPEN_TAG -> continue;
    $nextMeanigful = ...;
    if $nextMeanigful is not echo or print -> continue;
    if for complex option
    override
```

File: src/Fixer/PhpTag/ShortEchoTagFixer.php [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+
+    private function shortToLong(SplFileInfo $file, Tokens $tokens)
+    {
+        if (self::LONG_FUNCTION_PRINT === $this->configuration[self::OPTION_LONG_FUNCTION]) {
+            $echoToken = [\T_PRINT, 'print'];
+        } else {
+            $echoToken = [\T_ECHO, 'echo'];
+        }
+        $offset = $tokens->count() - 1;
+        for (;;) {
```
Comment (kubawerlos) [inline src/Fixer/PhpTag/ShortEchoTagFixer.php]: A classic loop would be more explicit or at least merge `$offset` and `$found` - they could be simply one variable, preferably named `$index`.

File: src/Fixer/PhpTag/ShortEchoTagFixer.php [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+            $offset = $found - 1;
+            $replace = [new Token([\T_OPEN_TAG, '<?php ']), new Token($echoToken)];
+            if (!$tokens[$found + 1]->isWhitespace()) {
+                $replace[] = new Token([\T_WHITESPACE, ' ']);
+            }
+            $tokens->overrideRange($found, $found, $replace);
+        }
+    }
+
+    private function isComplexCode(Tokens $tokens, $index)
```
Comment (kubawerlos) [inline src/Fixer/PhpTag/ShortEchoTagFixer.php]: I'm not sure id I understand it - can you add few more tests for this option?

File: src/Fixer/PhpTag/ShortEchoTagFixer.php [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+                $semicolonFound = true;
+            } elseif ($semicolonFound && !$token->isWhitespace()) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
```
Comment (kubawerlos) [inline src/Fixer/PhpTag/ShortEchoTagFixer.php]: No need for FQCN in this PHPDoc (BTW seems like an idea for new fixer).

File: src/Fixer/PhpTag/ShortEchoTagFixer.php [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+     * @param int                          $openTagIndex
+     * @param int                          $echoTagIndex
+     *
+     * @return \PhpCsFixer\Tokenizer\Token[]
+     */
+    private function buildLongToShortTokens(Tokens $tokens, $openTagIndex, $echoTagIndex)
+    {
+        $result = [new Token([\T_OPEN_TAG_WITH_ECHO, '<?='])];
+
+        // Skip to the first non-whitespace token between $openTagIndex and $echoTagIndex
```
Comment (kubawerlos) [inline src/Fixer/PhpTag/ShortEchoTagFixer.php]: From description it looks like `Tokens::getNextNonWhitespace`.

File: src/Fixer/PhpTag/ShortEchoTagFixer.php [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+                $found = $foundPrint;
+            }
+            $startRange = key($found);
+            $offset = $startRange + 1;
+            if (!$this->configuration[self::OPTION_SHORT_ALWAYS] && $this->isComplexCode($tokens, $offset)) {
+                continue;
+            }
+            next($found);
+            $endRange = key($found);
+            $newTokens = $this->buildLongToShortTokens($tokens, $startRange, $endRange);
```
Comment (kubawerlos) [inline src/Fixer/PhpTag/ShortEchoTagFixer.php]: Can't we simply remove the `echo`/`print` token with the following whitespace - if any - and override the tag open?

File: tests/Fixer/PhpTag/NoShortEchoTagFixerTest.php [hunk 13]
```diff
@@ -31,7 +31,7 @@ final class NoShortEchoTagFixerTest extends AbstractFixerTestCase
      */
     public function testFix($expected, $input = null)
     {
-        if (!ini_get('short_open_tag')) {
+        if (\PHP_VERSION_ID < 50400 && !ini_get('short_open_tag')) {
```
Comment (kubawerlos) [inline tests/Fixer/PhpTag/NoShortEchoTagFixerTest.php]: We don't support PHP version lower than `5.4`, so this change is pointless.

File: README.rst (hunk 2 above)
Comment (mlocati) [reply README.rst]: :+1: 94dc2c3dd377c5c08cd7c04c91a5ec1864bb20c5

File: dev-tools/trigger-website.sh (hunk 3 above)
Comment (mlocati) [reply dev-tools/trigger-website.sh]: Why? I agree that it's not strictly related to this PR, but IMHO the code is more clear...

For example, which of the following PHP code examples do you prefer?
```php
if (!($a === false)) {
```
vs
```php
if ($a !== false)) {
```

Furthermore, it fixes [this warning](https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/jobs/568483121#L633)

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 5 above)
Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: If there's a faster way to check if `$tokens` contains `<?php echo` or `<?php print` I'd be glad to update the code...

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 11 above)
Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: Not exactly the same: see the `between $openTagIndex and $echoTagIndex` part of the description

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 12 above)
Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: It's not that simple. For example:
```
<? /* comment */ echo 1 ?>
```
should become
```
<?= /* comment */ 1 ?>
```

(it's included in the fixer tests).

File: src/Fixer/PhpTag/NoShortEchoTagFixer.php (hunk 4 above)
Comment (mlocati) [reply src/Fixer/PhpTag/NoShortEchoTagFixer.php]: Done: 8ec8c60451e2e31844b1c0f81c1627b759c716ed

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 6 above)
Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: Done: f7a7d1a345ad755458c40708a79ec650141f59ef

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 9 above)
Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: There are already tests for it. BTW I added a PHPDoc comment to this method: dadc11e152f13bb15854f57ce4dc5eccf8c99f13

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 10 above)
Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: I do prefer FQCN in PHPDocs, and that's for these reasons:

1. If we need to reference a class only in PHPDocs, do we need to add a `use` statement at the beginning of the file?
If yes, it's quite useless to me.
If no: what if we add at a later time a reference to it in the code: should we fix all the previous PHPDocs?

2. consider this example:
    ```
    <?php
    use Namespace1\ClassName;
    use Namespace2\ClassName as ClassNameAlias;
    
    /**
     * @return ClassNameAlias
     */
    function foo() {
    }
    ```
    In order to know what `ClassNameAlias` actually is, humans have to read the `use` statements at the top of the file.

BTW, since this project doesn't use FQCN in comments, I updated this class: 9031768b641b9fb3ae82ea0b48053d0fd28cd5fe

File: tests/Fixer/PhpTag/NoShortEchoTagFixerTest.php (hunk 13 above)
Comment (mlocati) [reply tests/Fixer/PhpTag/NoShortEchoTagFixerTest.php]: Done: c3535e6f59568504b9a81418418a2194af780bcc

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 7 above)
Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: Done: 5b3d5eb6ce6373aa1cfdf73752814061f7913ed7

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 8 above)
Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: I simplified the code a bit: see 7feb5bee336c2c6ee6551f9ec47df45e5032b2cb

File: dev-tools/trigger-website.sh (hunk 3 above)
Comment (kubawerlos) [reply dev-tools/trigger-website.sh]: Sorry, I was thinking it's some leftover after old PRs.

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 5 above)
Comment (kubawerlos) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: We don't have to search for the sequence - see the `isCandidate` method in other fixers - it could use `findSequence` in few of them, but the goal is the method is fast, not super accurate - I'm not saying this is right (maybe more complex `isCandidate` could save some files from being checked) - but it is the approach all other fixers follow.

Simple `$tokens->isAnyTokenKindsFound([T_ECHO, T_PRINT])` would be fine, I guess.

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 12 above)
Comment (kubawerlos) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: I don't understand - that's exactly what I'm suggesting - `echo`  and ` ` after it are removed and `<?' becomes `<?=`.

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 11 above)
Comment (kubawerlos) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: Can we have a test for this "not exactly"? All tests are passing for https://github.com/mlocati/PHP-CS-Fixer/pull/1

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 5 above)
Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: You are absolutely right: 372f92e0f07fcb61ba78b83c4b5c918e4ae1ca69

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 12 above)
Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: Sorry, I didn't understand what you meant.

About removing the whitespace, we have two options: remove the whitespaces after the PHP opening tag, or remove the whitespaces after the `echo`/`print`.

Consider these two cases:

```
<?php /*comment*/ echo
  1
?>

<?php
  /*comment*/ echo 1
?>
```

what should be the short-echo version of them?

File: src/Fixer/PhpTag/ShortEchoTagFixer.php (hunk 11 above)
Comment (mlocati) [reply src/Fixer/PhpTag/ShortEchoTagFixer.php]: Yuu are right :+1:. That's because for sure the next non white-space token can only be an integer whose value is at max `$echoTagIndex`, it's better to use `getNextNonWhitespace()`

File: tests/Fixer/PhpTag/NoShortEchoTagFixerTest.php (hunk 13 above)
Comment (julienfalque) [reply tests/Fixer/PhpTag/NoShortEchoTagFixerTest.php]: The `if` can be completely removed here too.

Comment (julienfalque) [reply tests/Fixer/PhpTag/NoShortEchoTagFixerTest.php]: The `if` can be completely removed here too.

File: README.rst [hunk 14]
```diff
@@ -1799,6 +1800,20 @@ Choose from the list of available rules:
 
   *Risky rule: risky when the ``settype`` function is overridden or when used as the 2nd or 3rd expression in a ``for`` loop .*
 
+* **short_echo_tag** [@PhpCsFixer]
```
[the remark this answers is not in the document]
Comment (mlocati) [reply README.rst]: Done: ac53c2c0ab452cd9f2ecf8fce6f2fa6b07e66231

File: tests/Fixer/PhpTag/NoShortEchoTagFixerTest.php (hunk 13 above)
Comment (mlocati) [reply tests/Fixer/PhpTag/NoShortEchoTagFixerTest.php]: Done: 2721a030d64f2f00ef3952a5077857b545609a25

File: README.rst [hunk 15]
```diff
@@ -1799,6 +1800,20 @@ Choose from the list of available rules:
...
+
+  Replace short-echo ``<?=`` with long format ``<?php echo``/``<?php print``
```
[the remark this answers is not in the document]
Comment (mlocati) [reply README.rst]: Done: f986d5f05950fc1d832bdbe0775615dc45864241

File: src/RuleSet.php [hunk 16]
```diff
@@ -255,6 +254,7 @@ final class RuleSet implements RuleSetInterface
             'phpdoc_var_annotation_correct_order' => true,
             'protected_to_private' => true,
             'return_assignment' => true,
+            'echo_tag_syntax' => true,
```
Comment (julienfalque) [inline src/RuleSet.php]: Should be moved after `compact_nullable_typehint` for alphabetical order.

Comment (mlocati) [reply src/RuleSet.php]: Done: dbefb78ccb2a48ea59f31abc94cc0d2358b4718e

**Comment by mlocati** [comment]
There's still [this issue](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4493#discussion_r350040138) to be fixed.

How should these cases be translated from long-echo to short-echo tags?

1.
    ```php
    <?php /*comment*/ echo
      1
    ?>
    ```
2.
    ```php
    <?php
      /*comment*/ echo 1
    ?>
    ```
3.
    ```php
    <?php
      /*comment*/
      echo
      1
    ?>
    ```

**Comment by julienfalque** [comment]
IMO keeping current behavior is fine.

**Comment by mlocati** [comment]
> IMO keeping current behavior is fine.

Ok, so I'll update the code as per @kubawerlos [suggestion](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4493#discussion_r348714652).

**Comment by mlocati** [comment]
> Ok, so I'll update the code as per @kubawerlos [suggestion](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4493#discussion_r348714652).

Nope, I can't find a way to implement https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4493#discussion_r348714652 in a faster/simpler way.

So, unless @kubawerlos sees a way to implement it, IMHO this PR is ready to be merged...

**Comment by julienfalque** [comment]
Could you please rebase onto latest master to get your branch up-to-date?

**Comment by mlocati** [comment]
> Could you please rebase onto latest master to get your branch up-to-date?

Done.

File: src/Fixer/PhpTag/EchoTagSyntaxFixer.php [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+    /** @internal */
+    const FORMAT_LONG = 'long';
+
+    /** @internal */
+    const LONG_FUNCTION_ECHO = 'echo';
+
+    /** @internal */
+    const LONG_FUNCTION_PRINT = 'print';
+
+    private $supportedFormatOptions = [
```
Comment (kubawerlos) [inline src/Fixer/PhpTag/EchoTagSyntaxFixer.php]: Can be constant as well (same for `$supportedLongFunctionOptions`) - it is not changing during the run.

File: README.rst [hunk 18]
```diff
@@ -598,6 +598,20 @@ Choose from the list of available rules:
...
+  Replaces short-echo ``<?=`` with long format ``<?php echo``/``<?php print``
+  syntax, or vice-versa.
+
+  Configuration options:
+
+  - ``format`` (``'long'``, ``'short'``): the desired language construct; defaults to
+    ``'long'``
+  - ``long_function`` (``'echo'``, ``'print'``): the function to be used to expand the
+    short echo tags; defaults to ``'echo'``
+  - ``short_always`` (``bool``): always render short-echo tags even in case of
```
Comment (kubawerlos) [inline README.rst]: `handle_complex_code`? `short_always` looks meaningless.

Comment (mlocati) [reply README.rst]: `handle_complex_code` with a boolean value? I don't think it's very clear...

Comment (julienfalque) [reply README.rst]: What about `shorten_simple_statements_only`, defaulting to `true`? A bit verbose but very explicit :)

File: src/Fixer/PhpTag/EchoTagSyntaxFixer.php (hunk 17 above)
Comment (mlocati) [reply src/Fixer/PhpTag/EchoTagSyntaxFixer.php]: Done in https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4493/commits/034546b04580b786f795945c9f191dfc9cb8e218

File: README.rst (hunk 18 above)
Comment (mlocati) [reply README.rst]: Done in https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4493/commits/034546b04580b786f795945c9f191dfc9cb8e218

**Comment by julienfalque** [comment]
@mlocati Before merging this PR we should squash commits. As there is a commit from @kubawerlos, could you please squash all consecutive commits from same author?

**Comment by mlocati** [comment]
> Before merging this PR we should squash commits. As there is a commit from @kubawerlos, could you please squash all consecutive commits from same author?

Done.

**Comment by julienfalque** [comment]
Thank you @mlocati.

**Comment by mlocati** [comment]
So, I submitted this PR 16 months ago, it was reviewed 3 months later, and still today we don't have a release including it?
You should really redesign your release process guys, I think I made a huge mistake investing my time contributing to PHP-CS-Fixer.

**Comment by SpacePossum** [comment]
2.17 has been released with your rule. We are working on a new release model so we can release far more often.
Thank for all your work and for your patience!
